pkgs/test_api/lib/hooks_testing.dart

```diff
@@ -0,0 +1,130 @@
+/// the current [status], [result], and [errors] from the test.
+///
+/// Use [run] to run a test body and query for the success or failure.
+///
+/// Use [start] to start a test and query for whether it has finished running.
+class TestCaseMonitor {
+  final LiveTest _liveTest;
+  final _done = Completer<void>();
+  TestCaseMonitor._(FutureOr<void> Function() body)
+      : _liveTest = _createTest(body);
+
+  /// Run [body] as a test case and return a [TestCaseMonitor] with the result.
+  ///
+  /// The [status] of the returned result will always be [Status.complete].
+  /// The [result] and [errors] will reflect the latest state of the test.
+  /// {@template result-late-fail}
+  /// Note that a test can change result from success to failure even after it
+  /// has already completed if the test surfaces an unawaited asynchronous
+  /// error.
+  /// {@endtemplate}
+  ///
+  /// ```dart
+  /// final monitor = await TestCaseMonitor.run(() {
+  ///   fail('oh no!');
+  /// });
+  /// assert(monitor.result == Result.failure);
+  /// assert((monitor.errors.single.error as TestFailure).message == 'oh no!');
+  /// ```
+  static Future<TestCaseMonitor> run(FutureOr<void> Function() body) async {
+    final monitor = TestCaseMonitor.start(body);
+    await monitor.onDone;
+    return monitor;
+  }
+
+  /// Start [body] as a test case and return a [TestCaseMonitor] with the status
+  /// and result.
+  ///
+  /// The [status] of the test will be [Status.running] until it completes.
+  /// The [result] and [errors] will reflect the latest state of the test.
+  /// {@macro result-late-fail}
+  ///
+  /// ```dart
+  /// late void Function() completeWork;
+  /// final monitor = TestCaseMonitor.start(() {
+  ///   final outstandingWork = TestHandle.current.markPending();
+  ///   completeWork = outstandingWork.complete;
+  /// });
+  /// await pumpEventQueue();
+  /// assert(monitor.status == Status.running);
+  /// completeWork();
+  /// await monitor.onDone;
+  /// assert(monitor.status == Status.complete);
+  /// ```
+  /// The [result] and [errors] will reflect the latest state of the test.
+  static TestCaseMonitor start(FutureOr<void> Function() body) =>
+      TestCaseMonitor._(body).._start();
+
+  void _start() {
+    _liveTest.run().whenComplete(_done.complete);
+  }
+
+  /// A future that completes after this test has finished running, or has
+  /// surfaced an error.
+  Future<void> get onDone => _done.future;
+
+  /// The run status for the test.
+  Status get status => _liveTest.state.status;
+
+  /// The result for the test.
+  ///
+  /// A test that is still running may have a result of [Result.success] because
+  /// it has not failed _yet_. The result should only be read after the test is
+  /// done.
+  ///
+  /// {@macro result-late-fail}
+  ///
+  /// A failed test my be a [Result.failure] if the test failed with a
+  /// [TestFailure] exception, or a [Result.error] if it failed with any other
+  /// type of exception.
```

Comment:
I'm not sure I want to expose this... I'm tempted to create a new abstraction for this...

We want to drop the distinction, but have not gotten around to trying it and seeing if anything breaks.
https://github.com/dart-lang/test/issues/1465

I think it might be better to introduce this API with some other model instead of building more on top of it...